# it is a regration problem

## we are doing the machine learning on house price data set convert itno markdown

<table align="center" width=100%>
<tr>
    </td>
         </td>
     </td> 
     <td>
        <div align="center">
        <font color="#21618C" size=8px>
         <b> Regression modeling on House price Data 
        </b>
     </font>
    </div>
 </td>
</tr>
</table>
         

# Price Data 

# Problem Statement

#This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

# import 'pandas'  panda is utilize to format the data structure it can convert any data structure into pandas dataframe.
import pandas as pd
# import 'numpy'  to find stratestical summary we use numpy
import numpy as np

#import subpackage of Matplotlib
import matplotlib.pyplot as plt 
from matplotlib.colors import ListedColormap

#import 'Seaborn
import seaborn as sns

#to suppress warnings from warnings
import filtervarnings. filtersarnings('ignore')




### Load the dataset

In [9]:
# Load the & name it df_houseprice
import pandas as pd

In [10]:
df_houseprice = pd.read_csv('C:\\Users\\sai\\Downloads\\house_price.csv')

In [11]:
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

# Data overview

In [13]:
# Name the columns
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

# Interpretetion

 - Price :Price of the House.
 - resid_area :Residential Area where my house is located.

In [15]:
# basic info of the Dataset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


# Interpretation

- we have 13 numerical columns & 3 categorical column
- we have NULL values in the data
- we have total 16 columns in the data

In [18]:
# shape of the data 
df_houseprice.shape

(506, 16)

# interpretation

- we have total 16 columns
- we have 506 observations

In [21]:
# summarize the basic statistics of Data 
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

# interpretation

- we have outliers in data,most of the outliers are in AGE column

### data preprocessing

### a.data clinging

In [26]:
#show me first five observation
df_houseprice.head(5)

                   
                   

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [27]:
# rename  to the columns
#step 1 : columns (' show the columns names')
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [28]:
#step 2 : allocate new columns names to old columns names
# make the dicitionary('old columns nmae': 'new columns name')

d1={'price': 'Price',
'resid_aera': 'Residental_Area',
'air_quality':'Air_Quality',
'room_num': 'Room_Number',
'age': 'Age',
'distance': 'Distance',
'teachers': 'Teachers', 
'poor_prop': 'proportion_of_poor_population',
'airport': 'Airport', 
'n_hos_beds': 'Number_of_Hospital_Beds',
'n_hot_rooms': 'Tumber_of_Hotel_Rooms',
'waterbody':'Waterbody',
'rainfall': 'Rainfall',
'bus_ter': 'Bus_Terminal',
'Parks': 'Parks',
}




## b. Null Value Handling

In [30]:
df_houseprice=df_houseprice.rename(columns=d1)
df_houseprice.head(5)

Price  resid_area  air_qual  Room_Number   Age  Distance  Teachers  \
0   24.0       32.31     0.538        6.575  65.2      4.35      24.7   
1   21.6       37.07     0.469        6.421  78.9      4.99      22.2   
2   34.7       37.07     0.469        7.185  61.1      5.03      22.2   
3   33.4       32.18     0.458        6.998  45.8      6.21      21.3   
4   36.2       32.18     0.458        7.147  54.2      6.16      21.3   

   proportion_of_poor_population Airport  Number_of_Hospital_Beds  \
0                           4.98     YES                    5.480   
1                           9.14      NO                    7.332   
2                           4.03      NO                    7.394   
3                           2.94     YES                    9.268   
4                           5.33      NO                    8.824   

   Tumber_of_Hotel_Rooms Waterbody  Rainfall Bus_Terminal     parks  Sold  
0                11.1920     River        23          YES  0.049347     0  
1                12.1728      Lake        42          YES  0.046146     1  
2               101.1200       NaN        38          YES  0.045764     0  
3                11.2672      Lake        45          YES  0.047151     0  
4                11.2896      Lake        55          YES  0.039474     0

In [31]:
# Authenticate the data
df_houseprice.columns

Index(['Price', 'resid_area', 'air_qual', 'Room_Number', 'Age', 'Distance',
       'Teachers', 'proportion_of_poor_population', 'Airport',
       'Number_of_Hospital_Beds', 'Tumber_of_Hotel_Rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'parks', 'Sold'],
      dtype='object')

Laws that we need to followwhile processing the null values(specially for manufacturing)
- when we have less than 15% data is missing-> remove the observations
-betwwen 20-70% data is missing->impute the null values
-greater than 75% values are missing->remove the column

In [33]:
# find the value(null value)
df_houseprice.isnull().sum()

Price                              0
resid_area                         0
air_qual                           0
Room_Number                        0
Age                                0
Distance                           0
Teachers                           0
proportion_of_poor_population      0
Airport                            0
Number_of_Hospital_Beds            8
Tumber_of_Hotel_Rooms              0
Waterbody                        155
Rainfall                           0
Bus_Terminal                       0
parks                              0
Sold                               0
dtype: int64

In [34]:
# percentage missing values
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100,3)

Price                             0.000
resid_area                        0.000
air_qual                          0.000
Room_Number                       0.000
Age                               0.000
Distance                          0.000
Teachers                          0.000
proportion_of_poor_population     0.000
Airport                           0.000
Number_of_Hospital_Beds           1.581
Tumber_of_Hotel_Rooms             0.000
Waterbody                        30.632
Rainfall                          0.000
Bus_Terminal                      0.000
parks                             0.000
Sold                              0.000
dtype: float64

In [35]:
# find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   resid_area                     506 non-null    float64
 2   air_qual                       506 non-null    float64
 3   Room_Number                    506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   proportion_of_poor_population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_of_Hospital_Beds        498 non-null    float64
 10  Tumber_of_Hotel_Rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [36]:
#this is a categorical column so we can replace the null with the mode
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [37]:
#replace the null
df_houseprice.Waterbody=df_houseprice.Waterbody.fillna('River')

In [38]:
#percentage misssing values
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100,3)

Price                            0.000
resid_area                       0.000
air_qual                         0.000
Room_Number                      0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
proportion_of_poor_population    0.000
Airport                          0.000
Number_of_Hospital_Beds          1.581
Tumber_of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
parks                            0.000
Sold                             0.000
dtype: float64

In [39]:
#in Number_of_Hospital_Beds contain less than 10% missing values
df_houseprice.dropna(inplace=True)

In [40]:
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100.3)

Price                            0.0
resid_area                       0.0
air_qual                         0.0
Room_Number                      0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
proportion_of_poor_population    0.0
Airport                          0.0
Number_of_Hospital_Beds          0.0
Tumber_of_Hotel_Rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
parks                            0.0
Sold                             0.0
dtype: float64

In [41]:
#filter the numerical features in the dataset
#'select_dtype ' is used to select the variable with given data type
#' include=[np.number]' will include all categroial variables
#display categorial features
df_cat=df_houseprice.select_dtypes(include=["object"])
df_cat.columns



Index(['Airport', 'Waterbody', 'Bus_Terminal'], dtype='object')

In [42]:
#use 'get dummies' from pandas to create dummy variables
#use('drop_first' to create(n-1) dummy varibles
dummy_var = pd.get_dummies(data =df_cat , drop_first = True)
dummy_var = dummy_var.reset_index()

In [43]:
#use 'get dummies' from pandas to create dummy variables
#use('drop_first' to create(n-1) dummy varibles
dummy_var2 =pd.get_dummies(data = df_cat)

In [44]:
dummy_var

index  Airport_YES  Waterbody_Lake and River  Waterbody_River
0        0         True                     False             True
1        1        False                     False            False
2        2        False                     False             True
3        3         True                     False            False
4        4        False                     False            False
..     ...          ...                       ...              ...
493    501        False                      True            False
494    502         True                      True            False
495    503        False                     False             True
496    504         True                     False             True
497    505         True                     False             True

[498 rows x 4 columns]

In [45]:
import numpy as np

In [46]:
df_num = df_houseprice.select_dtypes(include = [np.number])
df_num.head()

Price  resid_area  air_qual  Room_Number   Age  Distance  Teachers  \
0   24.0       32.31     0.538        6.575  65.2      4.35      24.7   
1   21.6       37.07     0.469        6.421  78.9      4.99      22.2   
2   34.7       37.07     0.469        7.185  61.1      5.03      22.2   
3   33.4       32.18     0.458        6.998  45.8      6.21      21.3   
4   36.2       32.18     0.458        7.147  54.2      6.16      21.3   

   proportion_of_poor_population  Number_of_Hospital_Beds  \
0                           4.98                    5.480   
1                           9.14                    7.332   
2                           4.03                    7.394   
3                           2.94                    9.268   
4                           5.33                    8.824   

   Tumber_of_Hotel_Rooms  Rainfall     parks  Sold  
0                11.1920        23  0.049347     0  
1                12.1728        42  0.046146     1  
2               101.1200        38  0.045764     0  
3                11.2672        45  0.047151     0  
4                11.2896        55  0.039474     0

In [47]:
from sklearn.preprocessing import StandardScaler

In [48]:
df_target = df_houseprice['Price']

df_num =df_num.drop('Price' , axis =1)

In [49]:
#iniciate  the standard scaler
ss = StandardScaler()


In [50]:
#tranform the data
scale = ss.fit_transform(df_num)

In [51]:
scale

array([[-1.28244435, -0.13709359,  0.40897478, ..., -1.30206786,
        -0.47313205, -0.91152748],
       [-0.59005882, -0.73319492,  0.18997385, ...,  0.21519397,
        -0.77522463,  1.09705963],
       [-0.59005882, -0.73319492,  1.276446  , ..., -0.10422957,
        -0.81123666, -0.91152748],
       ...,
       [ 0.11687263,  0.16527665,  0.97923045, ..., -0.66322077,
         0.30293265,  1.09705963],
       [ 0.11687263,  0.16527665,  0.72041117, ...,  0.6144734 ,
         0.59750894,  1.09705963],
       [ 0.11687263,  0.16527665, -0.36606097, ...,  0.45476163,
         0.56369225,  1.09705963]])

In [52]:
scale = pd.DataFrame(scale , columns= df_num.columns)
scale = scale.reset_index()


In [53]:
scale.head()

index  resid_area  air_qual  Room_Number       Age  Distance  Teachers  \
0      0   -1.282444 -0.137094     0.408975 -0.115644  0.172845  1.462167   
1      1   -0.590059 -0.733195     0.189974  0.369741  0.476100  0.304963   
2      2   -0.590059 -0.733195     1.276446 -0.260905  0.495053  0.304963   
3      3   -1.301354 -0.828226     1.010516 -0.802977  1.054179 -0.111631   
4      4   -1.301354 -0.828226     1.222407 -0.505369  1.030487 -0.111631   

   proportion_of_poor_population  Number_of_Hospital_Beds  \
0                      -1.066368                -1.640299   
1                      -0.484136                -0.384875   
2                      -1.199330                -0.342847   
3                      -1.351886                 0.927490   
4                      -1.017382                 0.626514   

   Tumber_of_Hotel_Rooms  Rainfall     parks      Sold  
0              -0.352111 -1.302068 -0.473132 -0.911527  
1              -0.166019  0.215194 -0.775225  1.097060  
2              16.710340 -0.104230 -0.811237 -0.911527  
3              -0.337843  0.454762 -0.680402 -0.911527  
4              -0.333593  1.253320 -1.404723 -0.911527

In [54]:
#make dataframe called independent variable
df_independent = pd.concat([scale , dummy_var] , axis = 1)

df_independent.head()


index  resid_area  air_qual  Room_Number       Age  Distance  Teachers  \
0      0   -1.282444 -0.137094     0.408975 -0.115644  0.172845  1.462167   
1      1   -0.590059 -0.733195     0.189974  0.369741  0.476100  0.304963   
2      2   -0.590059 -0.733195     1.276446 -0.260905  0.495053  0.304963   
3      3   -1.301354 -0.828226     1.010516 -0.802977  1.054179 -0.111631   
4      4   -1.301354 -0.828226     1.222407 -0.505369  1.030487 -0.111631   

   proportion_of_poor_population  Number_of_Hospital_Beds  \
0                      -1.066368                -1.640299   
1                      -0.484136                -0.384875   
2                      -1.199330                -0.342847   
3                      -1.351886                 0.927490   
4                      -1.017382                 0.626514   

   Tumber_of_Hotel_Rooms  Rainfall     parks      Sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

In [55]:
# split  for the  training and training



In [56]:
import statsmodels
import statsmodels.api as sm

df_independent = sm.add_constant(df_independent)


In [57]:
df_independent

const  index  resid_area  air_qual  Room_Number       Age  Distance  \
0      1.0      0   -1.282444 -0.137094     0.408975 -0.115644  0.172845   
1      1.0      1   -0.590059 -0.733195     0.189974  0.369741  0.476100   
2      1.0      2   -0.590059 -0.733195     1.276446 -0.260905  0.495053   
3      1.0      3   -1.301354 -0.828226     1.010516 -0.802977  1.054179   
4      1.0      4   -1.301354 -0.828226     1.222407 -0.505369  1.030487   
..     ...    ...         ...       ...          ...       ...       ...   
493    1.0    493    0.116873  0.165277     0.434572  0.022531 -0.637413   
494    1.0    494    0.116873  0.165277    -0.238073  0.291796 -0.732180   
495    1.0    495    0.116873  0.165277     0.979230  0.798438 -0.779564   
496    1.0    496    0.116873  0.165277     0.720411  0.738208 -0.684797   
497    1.0    497    0.116873  0.165277    -0.366061  0.437057 -0.599506   

     Teachers  proportion_of_poor_population  Number_of_Hospital_Beds  \
0    1.462167                      -1.066368                -1.640299   
1    0.304963                      -0.484136                -0.384875   
2    0.304963                      -1.199330                -0.342847   
3   -0.111631                      -1.351886                 0.927490   
4   -0.111631                      -1.017382                 0.626514   
..        ...                            ...                      ...   
493 -1.176258                      -0.409958                 0.981720   
494 -1.176258                      -0.492534                -0.872945   
495 -1.176258                      -0.973995                -1.641654   
496 -1.176258                      -0.856429                 0.027273   
497 -1.176258                      -0.660486                 1.613499   

     Tumber_of_Hotel_Rooms  Rainfall     parks      Sold  index  Airport_YES  \
0                -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1                -0.166019  0.215194 -0.775225  1.097060      1        False   
2                16.710340 -0.104230 -0.811237 -0.911527      2        False   
3                -0.337843  0.454762 -0.680402 -0.911527      3         True   
4                -0.333593  1.253320 -1.404723 -0.911527      4        False   
..                     ...       ...       ...       ...    ...          ...   
493              -0.164805 -0.982644  0.155116  1.097060    501        False   
494               0.022197 -1.541636  0.522886  1.097060    502         True   
495              -0.162528 -0.663221  0.302933  1.097060    503        False   
496               0.403791  0.614473  0.597509  1.097060    504         True   
497              -0.549435  0.454762  0.563692  1.097060    505         True   

     Waterbody_Lake and River  Waterbody_River  
0                       False             True  
1                       False            False  
2                       False             True  
3                       False            False  
4                       False            False  
..                        ...              ...  
493                      True            False  
494                      True            False  
495                     False             True  
496                     False             True  
497                     False             True  

[498 rows x 18 columns]

In [58]:
from sklearn.model_selection import train_test_split


In [59]:
x_train, x_test, y_train , y_test = train_test_split(df_independent ,df_target, random_state=10, test_size = 0.2)                                                                                                 

In [60]:
print('x_train' , x_train.shape)
print('y_train' , y_train.shape)

print('x_test' , x_test.shape)
print('x_test' , y_test.shape)

x_train (398, 18)
y_train (398,)
x_test (100, 18)
x_test (100,)


# linear Regression


In [62]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [63]:
#create linear regression
model = LinearRegression()

#fit the model to the training data 
model.fit(x_train , y_train)

#predict on the testing data
prediction = model.predict(x_test)

In [67]:
prediction

array([22.54643539, 27.39840735,  6.01371194, 24.62279556, 12.15893075,
       13.24147688, 16.67022346, 25.66565456,  9.50703345, 28.11704874,
       20.10380925, 31.97996861, 27.7494175 , 21.2841941 , 26.64068987,
        8.1693733 , 23.29099208, 20.32076214, 35.32087816, 33.05755077,
       37.53089866, 22.31822842, 17.90935745, 24.93288423,  3.61313364,
       22.42015712, 24.04955853,  1.14080403, 22.1528586 , 21.98095176,
       30.99864887, 18.47460094, 24.47115012, 15.94716737, 24.85276297,
       27.60021664, 23.10156351, 23.57407559, 36.07856382, 41.83009608,
       30.57095683, 12.59562389, 19.78670774, 31.98447659, 13.3137158 ,
       26.24526278, 31.66815717, 24.94906032, 19.30734382, 24.98848122,
       23.56415543, 25.95215834, 26.68869766, 23.9541815 , 21.12523709,
       25.24908744, 10.13721903, 15.01905518, 20.7488436 , 25.20136288,
       19.53847673, 26.30949469, 33.61883399, 23.20007772, 34.05337745,
       15.58818707, 22.34851113, 18.35483462, 33.93630801, 33.10

In [71]:
mse = mean_squared_error(y_test , prediction)
print("mean_squared_error:" , mse)

mean_squared_error: 15.49818238426037


In [91]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test , prediction)
print("mean_absolute_error :", mae)

mean_absolute_error : 3.133142913286599
